akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessCoordinator.scala

```diff
@@ -0,0 +1,407 @@
+        }
+
+      case InternalGetResponse(_ @Replicator.NotFound(`key`)) =>
+        context.log.debug("{}: No previous state stored for Sharded Daemon Process", daemonProcessName)
+        stash.unstashAll(idle(initialState.value))
+
+      case InternalGetResponse(_ @Replicator.GetFailure(`key`)) =>
+        context.log.info("{}: Failed fetching initial state for Sharded Daemon Process, retrying", daemonProcessName)
+        start()
+
+      case other =>
+        stash.stash(other)
+        Behaviors.same
+    }
+  }
+
+  private def idle(currentState: ShardedDaemonProcessState): Behavior[ShardedDaemonProcessCommand] = {
+
+    Behaviors.receiveMessagePartial {
+      case request: ChangeNumberOfProcesses =>
+        if (request.newNumberOfProcesses == currentState.numberOfProcesses) {
+          context.log.info(
+            "{}: Rescale command for Sharded Daemon Process ignored, already running [{}] processes",
+            daemonProcessName,
+            request.newNumberOfProcesses)
+          request.replyTo ! StatusReply.Ack
+          Behaviors.same
+        } else {
+          val newState =
+            ShardedDaemonProcessState(
+              currentState.revision + 1,
+              request.newNumberOfProcesses,
+              completed = false,
+              started = Instant.now())
+          context.log.infoN(
+            "{}: Starting rescaling of Sharded Daemon Process to [{}] processes, rev [{}]",
+            daemonProcessName,
+            request.newNumberOfProcesses,
+            newState.revision)
+          prepareRescale(newState, request, currentState.numberOfProcesses)
+        }
+    }
+  }
+
+  private def prepareRescale(
+      newState: ShardedDaemonProcessState,
+      request: ChangeNumberOfProcesses,
+      previousNumberOfProcesses: Int): Behavior[ShardedDaemonProcessCommand] = {
+    replicatorAdapter.askUpdate(
+      replyTo =>
+        Replicator.Update(key, initialState, Replicator.WriteMajority(settings.rescaleWriteStateTimeout), replyTo)(
+          (register: Register) => register.withValue(selfUniqueAddress, newState)),
+      response => InternalUpdateResponse(response))
+
+    // wait for update to complete before applying
+    receiveWhileRescaling("prepareRescale") {
+      case InternalUpdateResponse(_ @Replicator.UpdateSuccess(`key`)) =>
+        pauseAllPingers(newState, Some(request), previousNumberOfProcesses)
+
+      case InternalUpdateResponse(_ @Replicator.UpdateTimeout(`key`)) =>
+        // retry
+        // FIXME should we count retries and eventually give up?
+        context.log.debug("{}: Failed updating scale state for sharded daemon process, retrying", daemonProcessName)
+        prepareRescale(newState, request, previousNumberOfProcesses)
+    }
+  }
+
+  // first step of rescale, pause all pinger actors so they do not wake the entities up again
+  // when we stop them
+  private def pauseAllPingers(
```

Comment:
One option we have is to move the pinging to the coordinator. Might simplify the process. Drawback is less ping redundancy in case of network partitions.

Reply:
Mmm, that would certainly simplify things quite a bit, I think it is an attractive idea.

Worst case scenario is worker crashes in face of partition, and isn't restarted until partition heals, but then you can work around that by making sure the worker doesn't crash by supervise.restart it.

Reply:
Done in 034116a